### PR TITLE
PyTorch DataLoader

### PR DESCRIPTION
**Pull request by atolopko-czi**
First draft of the PyTorch DataLoader support, sufficiently developed to warrant feedback from the community. Supports:
* User-specified obs/var queries for slicing the Census
* User-specified obs attributes to be returned in PyTorch Tensors (for training input or predicted labels)
* User-specified PyTorch batch size for items returned by DataLoader
* Automatic categorical encoding of obs attributes, with access to encoders for decoding, as needed
* Option for sparse or dense Tensors for X data
* Use of multiprocessing (multiple worker processes) to retrieve data in per-worker partitions
* Support splitting data into train/test/etc sets using PyTorch Data Pipes (e.g. RandomSplitter)
* Support shuffling data using PyTorch Data Pipes (e.g. Shuffler)
* Introduces an `experimental` sub-package, under which the new pytorch.py module exists (further sub-pkgs could be used)
* Demo code is in `api/python/notebooks/ml_demo/pytorch_lr_classifier.py`, which still needs to be converted into a proper notebook.
* Supports optional installation of pip packages for "experimental" package use. Specify as `pip install '.[experimental]'`.

Addresses:
- chanzuckerberg/single-cell-data-portal#478 
- chanzuckerberg/single-cell-data-portal#489
- chanzuckerberg/single-cell-data-portal#480 

Partially addresses:
- chanzuckerberg/single-cell-data-portal#476
- chanzuckerberg/single-cell-data-portal#477 
- chanzuckerberg/cellxgene#1762 
- chanzuckerberg/single-cell-infra#150

### CODE REVIEW

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pytorch.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+        tiledb_config={
+            "py.init_buffer_bytes": buffer_bytes,
+            "soma.init_buffer_bytes": buffer_bytes,
+        }
+    )
+    return soma.Experiment.open(uri, context=context)
+
+
+class _ObsAndXIterator(Iterator[ObsDatum]):
+    """
```
Comment (ebezzi) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/pytorch.py]: Probably not critical for an experimental module, but this docstring should use [Google-style guidelines](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html), which will also look neater and generate a proper docsite page.

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pytorch.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+            measurement_name=self.ms_name,
+            obs_query=self.obs_query,
+            var_query=self.var_query,
+        )
+
+        obs_joinids = self._query.obs_joinids()
+
+        worker_info = torch.utils.data.get_worker_info()
+        if worker_info:
+            partition, partitions = worker_info.id, worker_info.num_workers
```
Comment (ebezzi) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/pytorch.py]: `num_partitions`?

File: api/python/cellxgene_census/src/cellxgene_census/experimental/pytorch.py (hunk 1 above)
Comment (atolopko-czi) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/pytorch.py]: That reminds me, I need to add a "documentation" issue. Will clean it all up there!

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+    _obs_and_x_iter: Optional[_ObsAndXIterator]
+
+    _encoders: Optional[Encoders]
+
+    _stats: Stats
+
+    def __init__(
+        self,
+        experiment: soma.Experiment,
+        ms_name: str,
```
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:221]: lets make the names consistent with the rest of the cellxgene_census package, e.g., `get_anndata` uses `measurement_name`, not `ms_name`

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+ObsDatum = Tuple[Tensor, torch.sparse_coo_tensor]
+
+Encoders = Dict[str, LabelEncoder]
+
+DEFAULT_BUFFER_BYTES = 1024**3
+
+DEFAULT_WORKER_TIMEOUT = 120
+
+pytorch_logger = logging.getLogger("cellxgene_census.experimental.pytorch")
```
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:35]: I like that you are using the logger.  Should we expand this to other parts of the code base?

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+        var_joinids: pa.Array,
+        exp_uri: str,
+        aws_region: Optional[str],
+        measurement_name: str,
+        X_layer_name: str,
+        batch_size: int,
+        encoders: Dict[str, LabelEncoder],
+        stats: Stats,
+        obs_column_names: Sequence[str],
+        sparse_X: bool,
```
[the remark this answers is not in the document]
Comment (atarashansky) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:102]: Nevermind, I finally grokked how you're using this. Didn't know pytorch supporteed sparse data structures!!

File: api/python/cellxgene_census/pyproject.toml [hunk 6]
```diff
@@ -42,6 +42,12 @@ dependencies= [
     "certifi",
 ]
 
+[project.optional-dependencies]
+experimental = [
+    "torch==2.0.1",
```
Comment (bkmartinjr) [inline api/python/cellxgene_census/pyproject.toml:47]: missing sklearn?

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+            return
+
+        # TODO: support multiple layers, per somacore.query.query.ExperimentAxisQuery.to_anndata()
+        # TODO: for dev-time use, specify smaller batch size via platform_config, once supported
+
+        # TODO: handle closing of exp when iterator is no longer in use; may need be used as a ContextManager,
+        #  but not clear how we can do that when used by DataLoader
+        exp = _open_experiment(self.exp_uri, self.aws_region, buffer_bytes=self.buffer_bytes)
+
+        self._query = exp.axis_query(
```
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:268]: both experiment and query have a thread pool which may leak if not closed.  Query tries to avoid this via __del__ but that is unreliable.  if there is any cleanup protocol with pytorch, would be good to hook into it....

**Review by bkmartinjr — CHANGES_REQUESTED** [review]
Just small comments to resolve - overall it looks like a solid foundation

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py (hunk 7 above)
Comment (atolopko-czi) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:281]: Thanks, yeah, I've updated an earlier TODO to capture this: 
```
# TODO: handle closing of `_query` (and transitively, `exp`) when iterator is no longer in use; may need be used as a ContextManager, 
# but not clear how we can do that when used by DataLoader
```

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py (hunk 4 above)
Comment (bkmartinjr) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:31]: suggest a new issue for future debt work.  The builder also has lots of sloppy use of logger (my fault!)

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py (hunk 3 above)
Comment (bkmartinjr) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py]: To be clear:  I'm not suggesting you change the signature on this method/class.  Rather that a "sugar" factory function may be required.  this is entirely @pablo-gar 's call (and could wait for a future PR IMHO).

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py [hunk 8]
```diff
@@ -196,16 +196,38 @@ def obs_batch(self) -> pd.DataFrame:
...
         self.stats.n_soma_batches += 1
+        # TODO: also show per-batch stats (non-cumulative)
         pytorch_logger.debug(f"Retrieved batch: shape={self.X_batch.shape}, cum_stats: {self.stats}")
         return self.obs_batch_
 
 
 class ExperimentDataPipe(pipes.IterDataPipe[Dataset[ObsDatum]]):  # type: ignore
     """
-    An iterable-style data pipe.
+    An iterable-style PyTorch data pipe that reads obs and X data from a SOMA Experiment, and returns an iterator of
```
Comment (bkmartinjr) [inline api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:206]: I know you consider them redundant, but I still think these need lifecycle tags.  As Brian said, "belt and suspenders"

**Review by bkmartinjr — APPROVED** [review]
This is getting close enough for an experimental feature - approving.  Left a few additional comments you may want to consider.

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py (hunk 4 above)
Comment (atolopko-czi) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py:31]: https://github.com/chanzuckerberg/cellxgene-census/issues/504

File: api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py (hunk 3 above)
Comment (atolopko-czi) [reply api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py]: got it! update the TODO to reflect

File: api/python/notebooks/ml_demo/pytorch_lr_classifier.py
[the remark this answers is not in the document]
Comment (atolopko-czi) [reply api/python/notebooks/ml_demo/pytorch_lr_classifier.py:1]: I'm likely going to merge this shortly, but if you have any feedback you can add to this issue: https://github.com/chanzuckerberg/cellxgene-census/issues/476
